crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/ops.rs

```diff
@@ -1,21 +1,161 @@
+                // If a partial batch of hashes was found for this bloom then adjust
+                // the remote blooms time window to the cursor and queue it for future processing.
+                remote_bloom.time.start = cursor;
+
+                // Queue this bloom using the unique id if there is one.
+                let id = state.ops_batch_queue.0.share_mut(|queue, _| {
+                    Ok(queue.push_back(queue_id, QueuedOps::Bloom(remote_bloom)))
+                })?;
+
+                // If there was no id then a new one is created from the push_back call.
+                queue_id = Some(id);
+
+                data
+            }
+        };
+
+        self.batch_missing_ops(state, missing_hashes, queue_id)
+            .await
+    }
+
+    /// Generate the next batch of missing ops.
+    pub(super) async fn next_missing_ops_batch(
+        &self,
+        state: RoundState,
+    ) -> KitsuneResult<Vec<ShardedGossipWire>> {
+        // Pop the next queued batch.
+        let next_batch = state
+            .ops_batch_queue
+            .0
+            .share_mut(|queue, _| Ok(queue.pop_front()))?;
+
+        match next_batch {
+            // The next batch is hashes, batch them into ops using the queue id.
+            Some((queue_id, QueuedOps::Hashes(missing_hashes))) => {
+                self.batch_missing_ops(state, missing_hashes, Some(queue_id))
+                    .await
+            }
+            // The next batch is a bloom so the hashes need to be fetched before
+            // fetching the hashes.
+            Some((queue_id, QueuedOps::Bloom(remote_bloom))) => {
+                self.incoming_ops(state, remote_bloom, Some(queue_id)).await
+            }
+            // Nothing is queued so this node is done.
+            None => Ok(vec![ShardedGossipWire::missing_ops(
+                Vec::with_capacity(0),
+                MissingOpsStatus::AllComplete as u8,
+            )]),
+        }
+    }
+
+    /// Fetch missing ops into the appropriate size chunks of
+    /// and batch for future processing if there is too much data.
+    async fn batch_missing_ops(
+        &self,
+        state: RoundState,
+        mut missing_hashes: Vec<Arc<KitsuneOpHash>>,
+        mut queue_id: Option<usize>,
+    ) -> KitsuneResult<Vec<ShardedGossipWire>> {
+        let num_missing = missing_hashes.len();
+        let mut gossip = Vec::new();
+
+        // Fetch the missing ops if there is any.
+        let missing_ops = if missing_hashes.is_empty() {
+            Vec::with_capacity(0)
+        } else {
+            self.evt_sender
+                .fetch_op_data(FetchOpDataEvt {
+                    space: self.space.clone(),
+                    op_hashes: missing_hashes.clone(),
+                })
+                .await
+                .map_err(KitsuneError::other)?
+        };
+
+        let got_len = missing_ops.len();
+
+        // If there is less ops then missing hashes the call was batched.
```

Comment:
why?

Reply:
Because they are limited by the fetch ops call